aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -790,36 +804,151 @@ Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
+      Tensor start_end = compressed_indices.narrow(0, index / blocksize[0], 2).cpu();
+      AT_DISPATCH_INDEX_TYPES(start_end.scalar_type(), "select", [&]() {
+        auto accessor = start_end.accessor<index_t, 1>();
+        start = accessor[0];
+        end = accessor[1];
+      });
+      indices = plain_indices.slice(0, start, end);
+      values = self.values().slice(0, start, end);
+      select_nnz = end - start;
+    } else {
+      dim_indices = at::where(plain_indices.eq(index / blocksize[0]))[0];
+      // Assume that dim_indices is a sorted sequence of non-negative
+      // distinct integers. Below we'll try to solve `dim_indices ==
+      // arange(start, stop, step)`. If the solution exists then the
+      // select will be a view operation also for the `dim !=
+      // fast_dim` case.
+      select_nnz = dim_indices.numel();
+      int64_t step = -1;
+      if (select_nnz == 0) {
+        // trivial solution
+        start = end = 0;
+        step = 1;
+      } else {
+        Tensor dim_indices_cpu = dim_indices.cpu();
+        AT_DISPATCH_INDEX_TYPES(dim_indices_cpu.scalar_type(), "select", [&]() {
+          auto accessor = dim_indices_cpu.accessor<index_t, 1>();
+          if (accessor[select_nnz-1] - accessor[0] == select_nnz - 1) {
+            // special solution, avoids scanning dim_indices content
+            start = accessor[0];
+            end = start + select_nnz;
+            step = 1;
+          } else if (select_nnz > 1) {
+            // general solution, must scan the dim_indices content
+            int64_t step_candidate = accessor[1] - accessor[0];
+            for (int i=2; i<select_nnz; i++) {
+              if (accessor[i] - accessor[i-1] != step_candidate) {
+                step_candidate = 0;  // no solution
+                break;
+              }
+            }
+            if (step_candidate != 0) {
+              // found a general solution:
+              start = accessor[0];
+              end = start + select_nnz * step_candidate;
+              step = step_candidate;
+            }
+          }
+        });
+      }
+      indices = at::_convert_indices_from_csr_to_coo(compressed_indices, plain_indices)
+        .select(0, 0);
+      if (start != -1) {
+        indices = indices.slice(0, start, end, step);
+        values = self.values().slice(0, start, end, step);
+      } else {
+        // select will be a copy operation due to index_select usage!
+        indices = indices
+          .index_select(0, dim_indices);
+        is_copy = true;
+        values = self.values().index_select(0, dim_indices);
+      }
+    }
+
+    AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(self.layout(), "select",
+        [&]() {},
+        [&]() {
+          Tensor subblock_indices = at::arange(0, blocksize[1], indices_options);
+          indices = indices.mul(blocksize[1])
+            .repeat_interleave(blocksize[1])
+            .add(subblock_indices.repeat(select_nnz));
```

Comment:
We can replace `repeat_interleave` and `repeat` with a combination of broadcasting (i.e. `unsqueeze` at proper dims), `expand` and `view`.

Reply:
Indeed, the above is equivalent to
```c++
indices = indices.mul(blocksize[1]).unsqueeze(1).add(subblock_indices.unsqueeze(0)).flatten(0, 1);
```
but is about 5x more performant in the case of a small test.